// quicktype/dist/cli/analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ga = require("universal-analytics");
const storage = require("./storage");
const ci = require("./ci");
const uuid = require("uuid/v4");
const GoogleAnalyticsID = "UA-102732788-5";
class NoAnalytics {
    pageview(_page) {
        // Pass
    }
    timing(_category, _variable, _time) {
        // Pass
    }
    event(_category, _action, _label, _value) {
        // Pass
    }
}
exports.NoAnalytics = NoAnalytics;
class GoogleAnalytics {
    constructor() {
        const ciInfo = ci.getProvider();
        const userId = ciInfo !== undefined ? ciInfo.userId : storage.get("userId", uuid());
        this.visitor = ga(GoogleAnalyticsID, userId);
        if (ciInfo !== undefined) {
            this.ciProvider = ciInfo.provider;
        }
    }
    set ciProvider(provider) {
        this.visitor.set("dimension1", provider);
    }
    pageview(page) {
        this.visitor.pageview(page).send();
    }
    timing(category, variable, time) {
        this.visitor.timing(category, variable, time).send();
    }
    event(category, action, label, value) {
        if (label !== undefined) {
            if (value !== undefined) {
                this.visitor.event(category, action, label, value).send();
            }
            else {
                this.visitor.event(category, action, label).send();
            }
        }
        else {
            this.visitor.event(category, action).send();
        }
    }
}
exports.GoogleAnalytics = GoogleAnalytics;
